File: Android/Connect/src/main/java/siminov/connect/resource/ResourceUtils.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.connect.resource;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.StringTokenizer;

import siminov.connect.Constants;
import siminov.connect.exception.ServiceException;
import siminov.connect.model.ServiceDescriptor;
import siminov.connect.model.ServiceDescriptor.Request;
import siminov.connect.model.ServiceDescriptor.Request.HeaderParameter;
import siminov.connect.model.ServiceDescriptor.Request.QueryParameter;
import siminov.connect.service.design.IService;
import siminov.core.exception.SiminovException;
import siminov.core.log.Log;
import siminov.core.model.IDescriptor;
import siminov.core.utils.ClassUtils;

/**
 * It is a resource utility class, it exposes APIs to get service dynamic resources
 */
public class ResourceUtils {

	/**
	 * It resolve dynamic resources
	 * @param resourceValue Value of resource
	 * @param descriptors 
	 * @return
	 * @throws ServiceException
	 */
	public static String resolve(final String resourceValue, final IDescriptor...descriptors) throws ServiceException {
		
		if(resourceValue == null) {
			return resourceValue;
		}
		
		if(resourceValue.contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFER_REFERENCE)) {

			//Find {}
			int openingCurlyBracketIndex = resourceValue.indexOf(Constants.RESOURCE_SPACE) + 1;
			
			int singleClosingCurlyBracketIndex = resourceValue.indexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET);
			int doubleClosingCurlyBracketIndex = resourceValue.indexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET + Constants.RESOURCE_CLOSE_CURLY_BRACKET);

			String resourceKey;
			
			if(doubleClosingCurlyBracketIndex != -1) {

				resourceKey = resourceValue.substring(openingCurlyBracketIndex, doubleClosingCurlyBracketIndex + 1);
				int slashIndex = resourceKey.lastIndexOf(Constants.RESOURCE_SLASH);

				//Find {-
				String resourceClass = resourceKey.substring(0, resourceKey.substring(0, slashIndex).lastIndexOf(Constants.RESOURCE_DOT));
				String resourceAPI = resourceKey.substring(resourceKey.substring(0, slashIndex).lastIndexOf(Constants.RESOURCE_DOT) + 1, resourceKey.substring(0, slashIndex).length());

				Collection<Class<?>> resourceAPIParameterTypes = new LinkedList<Class<?>>();
				Collection<String> resourceAPIParameters = new LinkedList<String>();
				
				
				//Find -}}
				String apiParameters = resourceKey.substring(slashIndex + 1, resourceKey.lastIndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET) + 1);
				
				//Resolve all API parameters
				StringTokenizer apiParameterTokenizer = new StringTokenizer(apiParameters, Constants.RESOURCE_COMMA);
				
				while(apiParameterTokenizer.hasMoreElements()) {
					String apiParameter = apiParameterTokenizer.nextToken();
					
					resourceAPIParameterTypes.add(String.class);
					resourceAPIParameters.add(resolve(apiParameter, descriptors));
				}
				
			
				int count = 0;
				Class<?>[] apiParameterTypes = new Class<?>[resourceAPIParameters.size()];
				for(Class<?> resourceAPIParameterType : resourceAPIParameterTypes) {
					apiParameterTypes[count++] = resourceAPIParameterType;
				}
				

				Object classObject = ClassUtils.createClassInstance(resourceClass);
				String resolvedValue = null;
				try {
					resolvedValue = (String) ClassUtils.invokeMethod(classObject, resourceAPI, apiParameterTypes, resourceAPIParameters.toArray());
				} catch(SiminovException se) {
					Log.error(ResourceUtils.class.getName(), "resolve", "SiminovException caught while invoking method, RESOURCE-API: " + resourceAPI + ", " + se.getMessage());
					throw new ServiceException(ResourceUtils.class.getName(), "resolve", se.getMessage());
				}

				
				return resolve(resolvedValue, descriptors);
			} else {

				resourceKey = resourceValue.substring(openingCurlyBracketIndex, singleClosingCurlyBracketIndex);
				int dotIndex = resourceKey.lastIndexOf(Constants.RESOURCE_DOT);

				String resourceClass = resourceKey.substring(0, dotIndex);

				String resourceAPI = resourceKey.substring(resourceKey.lastIndexOf(Constants.RESOURCE_DOT) + 1, resourceKey.length());
				
				Object classObject = ClassUtils.createClassInstance(resourceClass);
				String value = null;
				try {
					value = (String) ClassUtils.getValue(classObject, resourceAPI);
				} catch(SiminovException se) {
					Log.error(ResourceUtils.class.getName(), "resolve", "SiminovException caught while getting values, RESOURCE-API: " + resourceAPI + ", " + se.getMessage());
					throw new ServiceException(ResourceUtils.class.getName(), "resolve", se.getMessage());
				}

				
				String resolvedValue = resourceValue.replace(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFER_REFERENCE + Constants.RESOURCE_SPACE + resourceKey + Constants.RESOURCE_CLOSE_CURLY_BRACKET, value);
				return resolve(resolvedValue, descriptors);
			}
		} else if(resourceValue.contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_SELF_REFERENCE)) {
			
			String key = resourceValue.substring(resourceValue.indexOf(Constants.RESOURCE_SPACE) + 1, resourceValue.indexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET));
			String value = null;
			
			
			for(IDescriptor descriptor: descriptors) {
				
				if(descriptor.containProperty(key)) {
					value = descriptor.getProperty(key);
					break;
				}
			}
			
			return resolve(value, descriptors);
		} else if(resourceValue.contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE)) {
			
			String key = resourceValue.substring(resourceValue.indexOf(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE) + (Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE).length() + 1, resourceValue.indexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET));
			String value = null;
			
			for(IDescriptor descriptor: descriptors) {
				
				if(descriptor.containProperty(key)) {
					value = descriptor.getProperty(key);
					break;
				}
			}

			
			String resolvedValue = resourceValue.replace(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE + " " + key + Constants.RESOURCE_CLOSE_CURLY_BRACKET, value);
			return resolve(resolvedValue, descriptors);
		} 
		
		return resourceValue;
	}
	
	
	/**
	 * It resolve service descriptor instance 
	 * @param service Service instance
	 * @throws ServiceException If any exception occur while resolving dynamic resources
	 */
	public static void resolve(final IService service) throws ServiceException {

		/*
		 * Resolve Service Descriptor Properties
		 */
		ServiceDescriptor serviceDescriptor = service.getServiceDescriptor();
		
		Iterator<String> resources = service.getResources();
		while(resources.hasNext()) {
			String resourceName = resources.next();
			Object resourceValue = service.getResource(resourceName);
			
			if(resourceValue instanceof String) {
				serviceDescriptor.addProperty(resourceName, (String) resourceValue);
			}
		}
		
		
		Iterator<String> serviceDescriptorProperties = serviceDescriptor.getProperties();
		while(serviceDescriptorProperties.hasNext()) {
			
			String serviceDescriptorProperty = serviceDescriptorProperties.next();
			String serviceDescriptorValue = serviceDescriptor.getProperty(serviceDescriptorProperty);
			serviceDescriptorValue = ResourceUtils.resolve(serviceDescriptorValue, serviceDescriptor);
			
			serviceDescriptor.addProperty(serviceDescriptorProperty, serviceDescriptorValue);
		}
		
		
		/*
		 * Resolve API Properties
		 */
		Request request = serviceDescriptor.getRequest(service.getRequest());
		Iterator<String> apiProperties = request.getProperties();
		while(apiProperties.hasNext()) {
			
			String apiProperty = apiProperties.next();
			String apiValue = request.getProperty(apiProperty);
			apiValue = ResourceUtils.resolve(apiValue, serviceDescriptor, request);
			
			request.addProperty(apiProperty, apiValue);
		}

		
		/*
		 * Resolve API Query Parameters
		 */
		Iterator<QueryParameter> queryParameters = request.getQueryParameters();
		while(queryParameters.hasNext()) {
			
			QueryParameter queryParameter = queryParameters.next();
			
			String queryValue = queryParameter.getValue();
			queryValue = ResourceUtils.resolve(queryValue, serviceDescriptor, request, queryParameter);
			
			queryParameter.setValue(queryValue);
		}

		
		/*
		 * Resolve API Query Parameters
		 */
		Iterator<HeaderParameter> headerParameters = request.getHeaderParameters();
		while(headerParameters.hasNext()) {
			
			HeaderParameter headerParameter = headerParameters.next();
			
			String headerValue = headerParameter.getValue();
			headerValue = ResourceUtils.resolve(headerValue, serviceDescriptor, request, headerParameter);
			
			headerParameter.setValue(headerValue);
		}
	}
}

File: Windows/universal8.1/Siminov/Connect/Resource/ResourceUtils.cs
///
/// [SIMINOV FRAMEWORK - CONNECT]
/// Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.




using Siminov.Connect.Exception;
using Siminov.Connect.Model;
using Siminov.Connect.Service.Design;
using Siminov.Core.Exception;
using Siminov.Core.Log;
using Siminov.Core.Model;
using Siminov.Core.Utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Siminov.Connect.Resource
{


    /// <summary>
    /// It is a resource utility class, it exposes APIs to get service dynamic resources
    /// </summary>
    public class ResourceUtils
    {


        /// <summary>
        /// It resolve dynamic resources
        /// </summary>
        /// <param name="resourceValue">Value of resource</param>
        /// <param name="descriptors">Array of descriptors</param>
        /// <returns>ServiceException</returns>
	    public static String Resolve(String resourceValue, IDescriptor[] descriptors) 
        {
		
		    if(resourceValue == null) 
            {
			    return resourceValue;
		    }
		
		    if(resourceValue.Contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFER_REFERENCE)) 
            {

			    //Find {}
			    int openingCurlyBracketIndex = resourceValue.IndexOf(Constants.RESOURCE_SPACE) + 1;
			
			    int singleClosingCurlyBracketIndex = resourceValue.IndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET);
			    int doubleClosingCurlyBracketIndex = resourceValue.IndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET + Constants.RESOURCE_CLOSE_CURLY_BRACKET);

			    String resourceKey;
			
			    if(doubleClosingCurlyBracketIndex != -1) 
                {

				    resourceKey = resourceValue.Substring(openingCurlyBracketIndex, doubleClosingCurlyBracketIndex + 1);
				    int slashIndex = resourceKey.LastIndexOf(Constants.RESOURCE_SLASH);

				    //Find {-
				    String resourceClass = resourceKey.Substring(0, resourceKey.Substring(0, slashIndex).LastIndexOf(Constants.RESOURCE_DOT));
				    String resourceAPI = resourceKey.Substring(resourceKey.Substring(0, slashIndex).LastIndexOf(Constants.RESOURCE_DOT) + 1, resourceKey.Substring(0, slashIndex).Length);

				    ICollection<Type> resourceAPIParameterTypes = new LinkedList<Type>();
				    ICollection<String> resourceAPIParameters = new LinkedList<String>();
				
				
				    //Find -}}
				    String apiParameters = resourceKey.Substring(slashIndex + 1, resourceKey.LastIndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET) + 1);
				
				    //Resolve all API parameters
                    String[] apiParameterTokenizer = Regex.Split(apiParameters, Constants.RESOURCE_COMMA);

                    for (int i = 0; i < apiParameterTokenizer.Length;i++)
                    {
                        String apiParameter = apiParameterTokenizer[i];

                        resourceAPIParameterTypes.Add(typeof(String));
                        resourceAPIParameters.Add(Resolve(apiParameter, descriptors));
                    }
				
			
				    int count = 0;
				    Type[] apiParameterTypes = new Type[resourceAPIParameters.Count];
				    foreach(Type resourceAPIParameterType in resourceAPIParameterTypes) 
                    {
					    apiParameterTypes[count++] = resourceAPIParameterType;
				    }
				

				    Object classObject = ClassUtils.CreateClassInstance(resourceClass);
				    String resolvedValue = null;
				    try 
                    {
					    resolvedValue = (String) ClassUtils.InvokeMethod(classObject, resourceAPI, apiParameterTypes, resourceAPIParameters.ToArray());
				    } 
                    catch(SiminovException se) 
                    {
					    Log.Error(typeof(ResourceUtils).Name, "Resolve", "SiminovException caught while invoking method, RESOURCE-API: " + resourceAPI + ", " + se.Message);
					    throw new ServiceException(typeof(ResourceUtils).Name, "Resolve", se.GetMessage());
				    }

				
				    return Resolve(resolvedValue, descriptors);
			    } else {

				    resourceKey = resourceValue.Substring(openingCurlyBracketIndex, singleClosingCurlyBracketIndex);
				    int dotIndex = resourceKey.LastIndexOf(Constants.RESOURCE_DOT);

				    String resourceClass = resourceKey.Substring(0, dotIndex);

				    String resourceAPI = resourceKey.Substring(resourceKey.LastIndexOf(Constants.RESOURCE_DOT) + 1, resourceKey.Length);
				
				    Object classObject = ClassUtils.CreateClassInstance(resourceClass);
				    String value = null;
				    try 
                    {
					    value = (String) ClassUtils.GetValue(classObject, resourceAPI);
				    } 
                    catch(SiminovException se) 
                    {
					    Log.Error(typeof(ResourceUtils).Name, "resolve", "SiminovException caught while getting values, RESOURCE-API: " + resourceAPI + ", " + se.GetMessage());
					    throw new ServiceException(typeof(ResourceUtils).Name, "resolve", se.GetMessage());
				    }

				
				    String resolvedValue = resourceValue.Replace(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFER_REFERENCE + Constants.RESOURCE_SPACE + resourceKey + Constants.RESOURCE_CLOSE_CURLY_BRACKET, value);
				    return Resolve(resolvedValue, descriptors);
			    }
		    } 
            else if(resourceValue.Contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_SELF_REFERENCE)) 
            {
			
			    String key = resourceValue.Substring(resourceValue.IndexOf(Constants.RESOURCE_SPACE) + 1, resourceValue.IndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET));
			    String value = null;
			
			
			    foreach(IDescriptor descriptor in descriptors) {
				
				    if(descriptor.ContainProperty(key)) {
					    value = descriptor.GetProperty(key);
					    break;
				    }
			    }
			
			    return Resolve(value, descriptors);
		    } 
            else if(resourceValue.Contains(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE)) 
            {

                String key = resourceValue.Substring(resourceValue.IndexOf(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE) + (Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE).Length + 1, resourceValue.IndexOf(Constants.RESOURCE_CLOSE_CURLY_BRACKET) - (resourceValue.IndexOf(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE) + (Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE).Length + 1));
			    String value = null;
			
			    foreach(IDescriptor descriptor in descriptors) 
                {
				
				    if(descriptor.ContainProperty(key)) 
                    {
					    value = descriptor.GetProperty(key);
					    break;
				    }
			    }

			
			    String resolvedValue = resourceValue.Replace(Constants.RESOURCE_OPEN_CURLY_BRACKET + Constants.RESOURCE_REFERENCE + " " + key + Constants.RESOURCE_CLOSE_CURLY_BRACKET, value);
			    return Resolve(resolvedValue, descriptors);
		    } 
		
		    return resourceValue;
	    }
	
	

        /// <summary>
        /// It resolve service descriptor instance 
        /// </summary>
        /// <param name="service">Service instance</param>
        /// <exception cref="Siminov.Connect.Exception.ServiceException">If any exception occur while resolving dynamic resources</exception>
	    public static void Resolve(IService service) 
        {

		    /*
		     * Resolve Service Descriptor Properties
		     */
		    ServiceDescriptor serviceDescriptor = service.GetServiceDescriptor();
		
		    IEnumerator<String> resources = service.GetResources();
		    while(resources.MoveNext()) 
            {
			    String resourceName = resources.Current;
			    Object resourceValue = service.GetResource(resourceName);
			
			    if(resourceValue is String) 
                {
				    serviceDescriptor.AddProperty(resourceName, (String) resourceValue);
			    }
		    }
		
		
		    IEnumerator<String> serviceDescriptorProperties = serviceDescriptor.GetProperties();
            IDictionary<String, String> changedServiceDescriptorProperties = new Dictionary<String, String>();

		    while(serviceDescriptorProperties.MoveNext()) 
            {
			
			    String serviceDescriptorProperty = serviceDescriptorProperties.Current;
			    String serviceDescriptorValue = serviceDescriptor.GetProperty(serviceDescriptorProperty);

                var serviceDescriptors = new List<IDescriptor>();
                serviceDescriptors.Add(serviceDescriptor);
                
                serviceDescriptorValue = ResourceUtils.Resolve(serviceDescriptorValue, serviceDescriptors.ToArray());

                changedServiceDescriptorProperties.Add(serviceDescriptorProperty, serviceDescriptorValue);
		    }

            foreach (String changedServiceDescriptorProperty in changedServiceDescriptorProperties.Keys) 
            {
                serviceDescriptor.AddProperty(changedServiceDescriptorProperty, changedServiceDescriptorProperties[changedServiceDescriptorProperty]);
            }
		
		
		    /*
		     * Resolve API Properties
		     */
		    Connect.Model.ServiceDescriptor.Request request = serviceDescriptor.GetRequest(service.GetRequest());
		    IEnumerator<String> apiProperties = request.GetProperties();
            IDictionary<String, String> changedApiProperties = new Dictionary<String, String>();

            while(apiProperties.MoveNext()) 
            {
			
			    String apiProperty = apiProperties.Current;
			    String apiValue = request.GetProperty(apiProperty);

                var serviceDescriptors = new List<IDescriptor>();
                serviceDescriptors.Add(serviceDescriptor);

			    apiValue = ResourceUtils.Resolve(apiValue, serviceDescriptors.ToArray());

                changedApiProperties.Add(apiProperty, apiValue);
		    }

            foreach (String changedApiProperty in changedApiProperties.Keys)
            {
                request.AddProperty(changedApiProperty, changedApiProperties[changedApiProperty]);
            }
		
		    /*
		     * Resolve API Query Parameters
		     */
		    IEnumerator<Connect.Model.ServiceDescriptor.Request.QueryParameter> queryParameters = request.GetQueryParameters();
            while(queryParameters.MoveNext()) 
            {
			
			    Connect.Model.ServiceDescriptor.Request.QueryParameter queryParameter = queryParameters.Current;
			
			    String queryValue = queryParameter.GetValue();

                var serviceDescriptors = new List<IDescriptor>();
                serviceDescriptors.Add(serviceDescriptor);

			    queryValue = ResourceUtils.Resolve(queryValue, serviceDescriptors.ToArray());
			
			    queryParameter.SetValue(queryValue);
		    }

		
		    /*
		     * Resolve API Query Parameters
		     */
		    IEnumerator<Connect.Model.ServiceDescriptor.Request.HeaderParameter> headerParameters = request.GetHeaderParameters();
            while(headerParameters.MoveNext()) 
            {
			
			    Connect.Model.ServiceDescriptor.Request.HeaderParameter headerParameter = headerParameters.Current;
			
			    String headerValue = headerParameter.GetValue();

                var serviceDescriptors = new List<IDescriptor>();
                serviceDescriptors.Add(serviceDescriptor);

			    headerValue = ResourceUtils.Resolve(headerValue, serviceDescriptors.ToArray());
			
			    headerParameter.SetValue(headerValue);
		    }
	    }
    }
}
